### PR TITLE
kubeadm upgrade plan: print a component config state table

### PR DESCRIPTION
**Pull request by rosti**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

This change enables `kubeadm upgrade plan` to print a state table with information regarding known component config API groups. Most importantly this information includes current and preferred version for each group and an indication if a manual config upgrade is required.

In addition to that, this PR also replaces most of the versioned use of the output API in `kubeadm upgrade plan` with the internal types.

**Which issue(s) this PR fixes**:

Refs:
- kubernetes/kubeadm#1940
- kubernetes/enhancements#1381

**Special notes for your reviewer**:

This PR is part of the implementation of the new kubeadm component config management scheme KEP (see link below). It also depends on #91980  . Please, review the last 2 commits only!

/cc @kubernetes/sig-cluster-lifecycle-pr-reviews
/area kubeadm
/priority important-longterm
/assign @fabriziopandini @neolit123 @ereslibre
/hold

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm: upgrade plan now prints a table indicating the state of known component configs prior to upgrade
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/kubeadm/20190925-component-configs.md
```

### CODE REVIEW

**Comment by neolit123** [comment]
@rosti can you please show example output?

File: cmd/kubeadm/app/cmd/upgrade/common.go [hunk 1]
```diff
@@ -75,7 +75,7 @@ func getK8sVersionFromUserInput(flags *applyPlanFlags, args []string, versionIsM
 }
 
 // enforceRequirements verifies that it's okay to upgrade and then returns the variables needed for the rest of the procedure
-func enforceRequirements(flags *applyPlanFlags, dryRun bool, newK8sVersion string) (clientset.Interface, upgrade.VersionGetter, *kubeadmapi.InitConfiguration, error) {
+func enforceRequirements(plan bool, flags *applyPlanFlags, dryRun bool, newK8sVersion string) (clientset.Interface, upgrade.VersionGetter, *kubeadmapi.InitConfiguration, error) {
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/upgrade/common.go]: cc @Klaven

File: cmd/kubeadm/app/apis/kubeadm/types.go [hunk 2]
```diff
@@ -425,6 +425,9 @@ type ComponentConfig interface {
 	// DeepCopy should create a new deep copy of the component config in place
 	DeepCopy() ComponentConfig
 
+	// Version returns a Kubernetes API version of the component config (v1alpha1, v1beta1, v1, etc.)
+	Version() string
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/types.go]: i think we might want to have VersionKind() here instead even if our third-party CCs don't have multiple kinds yet.

Comment (rosti) [reply cmd/kubeadm/app/apis/kubeadm/types.go]: I don't want kinds to be visible at all (let alone at this level). What I was thinking, and I'll probably change it on an iteration is to have this as a `GroupVersion`.
The idea behind `ComponentConfig` objects is to have them hide the details around a given `GroupVersion`. This includes what kinds there are, what goes where and how to structure it.
This helps isolate version specific code. Plus, adding new component config versions becomes more easy.

**Comment by rosti** [comment]
Sample output:
```
# kubeadm upgrade plan --allow-experimental-upgrades
...
Components that must be upgraded manually after you have upgraded the control plane with 'kubeadm upgrade apply':
COMPONENT   CURRENT                                        AVAILABLE
Kubelet     1 x v1.18.0-alpha.2.584+f29ad2e1cc3596-dirty   v1.18.0-alpha.5

Upgrade to the latest experimental version:

COMPONENT            CURRENT                                    AVAILABLE
API Server           v1.18.0-alpha.2.584+f29ad2e1cc3596-dirty   v1.18.0-alpha.5
Controller Manager   v1.18.0-alpha.2.584+f29ad2e1cc3596-dirty   v1.18.0-alpha.5
Scheduler            v1.18.0-alpha.2.584+f29ad2e1cc3596-dirty   v1.18.0-alpha.5
Kube Proxy           v1.18.0-alpha.2.584+f29ad2e1cc3596-dirty   v1.18.0-alpha.5
CoreDNS              1.6.5                                      1.6.5
Etcd                 3.4.3                                      3.4.3-0

You can now apply the upgrade by executing the following command:

        kubeadm upgrade apply v1.18.0-alpha.5 --allow-experimental-upgrades

Note: Before you can perform this upgrade, you have to update kubeadm to v1.18.0-alpha.5.

_____________________________________________________________________


The table below shows the current state of component configs as understood by this version of kubeadm.
Configs that have a "yes" mark in the "USER ACTION REQUIRED" column require manual config upgrade or
resetting to kubeadm defaults before a successful upgrade can be performed. The version to manually
upgrade to is denoted in the "PREFERRED VERSION" column.

API GROUP                 CURRENT VERSION   PREFERRED VERSION   USER ACTION REQUIRED
kubeproxy.config.k8s.io   v1alpha1          v1alpha1            no
kubelet.config.k8s.io     v1beta1           v1beta1             no
_____________________________________________________________________
#
```

File: cmd/kubeadm/app/apis/kubeadm/types.go (hunk 2 above)
Comment (neolit123) [reply cmd/kubeadm/app/apis/kubeadm/types.go]: Groups are not supposed to change, but a Group can introduce new Kinds during upgrade to a new Version. not tracking new Kinds means no visibility for this aspect from "plan", correct?

GroupVersion-only tracking would mean that "plan" can say, hey i know there is a new version  for this Group, but i'm not going to tell you what changed in it.

Comment (neolit123) [reply cmd/kubeadm/app/apis/kubeadm/types.go:429]: after discussions on this, having `group | version` is fine.

**Review by neolit123 — COMMENTED** [review]
/hold cancel
we have a number of PRs in the queue and it might be quite packed this cycle.

**Pull request merged**